### PR TITLE
Users want to simulate lossy networks

### PR DESCRIPTION
**Pull request by jrw972**
Problem
-------

The OpenDDS user community is interested in running OpenDDS with RTPS
in lossy networks, e.g., mobile data networks.  Testing directly on
these networks or using a physical or virtual device to condition the
network is not always practical.

Solution
--------

Provide configuration options that cause messages to probabilistically
be dropped.

### CODE REVIEW

**Comment by jwillemsen** [comment]
Would this be something that needs to be explicitly enabled at compile time? I can imagine that people don't want this in their shipped application

File: dds/DCPS/transport/rtps_udp/RtpsUdpSendStrategy.cpp [hunk 1]
```diff
@@ -301,6 +304,20 @@ RtpsUdpSendStrategy::send_single_i(const iovec iov[], int n,
...
 
+#ifdef OPENDDS_TESTING_FEATURES
+  double m;
+  double b;
+  if (link_->transport().config().drop_messages(m, b)) {
+    size_t total_length = 0;
+    for (int i = 0; i < n; ++i) {
+      total_length += iov[i].iov_len;
+    }
+    if (drand48() < total_length * m + b) {
```
Comment (iguessthislldo) [inline dds/DCPS/transport/rtps_udp/RtpsUdpSendStrategy.cpp:315]: I saw this was having problems before. Looks like it's POSIX only.

File: MPC/config/dcps_optional_features.mpb [hunk 2]
```diff
@@ -59,5 +59,9 @@ feature (!no_opendds_security) : taoidldefaults, dcps_ts_defaults {
   dcps_ts_flags += -DOPENDDS_SECURITY
 }
 
+feature (!no_opendds_testing_features) {
```
Comment (jrw972) [inline MPC/config/dcps_optional_features.mpb:62]: Move to separate MPB and make core and rtps use it.

File: dds/DCPS/RTPS/Sedp.h [hunk 3]
```diff
@@ -655,6 +655,10 @@ class Sedp : public DCPS::RcEventHandler {
     type_lookup_service_ = type_lookup_service;
   }
 
+#ifdef OPENDDS_TESTING_FEATURES
```
Comment (jrw972) [inline dds/DCPS/RTPS/Sedp.h:658]: Remove #defines in headers.

File: dds/DCPS/transport/rtps_udp/RtpsUdpInst.h [hunk 4]
```diff
@@ -167,6 +208,11 @@ class OpenDDS_Rtps_Udp_Export RtpsUdpInst : public TransportInst {
   ACE_INET_Addr rtps_relay_address_;
   bool use_ice_;
   ACE_INET_Addr stun_server_address_;
+#ifdef OPENDDS_TESTING_FEATURES
```
Comment (jrw972) [inline dds/DCPS/transport/rtps_udp/RtpsUdpInst.h:211]: Move all of this to base class.

File: dds/DCPS/transport/rtps_udp/RtpsUdpSendStrategy.cpp [hunk 5]
```diff
@@ -31,6 +31,9 @@
 #endif
 
 #include <cstring>
+#ifdef OPENDDS_TESTING_FEATURES
```
Comment (jrw972) [inline dds/DCPS/transport/rtps_udp/RtpsUdpSendStrategy.cpp:34]: Check for others and eliminate where not necessary.

**Review by mitza-oci — APPROVED** [review]

**Pull request merged**